pm/recipient_test.go

```diff
@@ -0,0 +1,436 @@
+	sig := []byte("foo")
+
+	r := NewRecipient(b, v, ts, secret, faceValue, winProb)
+
+	params1, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero deposit and penalty escrow
+	b.SetDeposit(sender, big.NewInt(0))
+	b.SetPenaltyEscrow(sender, big.NewInt(0))
+
+	// Test zero deposit and penalty escrow error
+	ticket := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       0,
+		RecipientRandHash: params1.RecipientRandHash,
+	}
+
+	won, err := r.ReceiveTicket(ticket, sig, params1.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fail()
+	}
+
+	err = r.RedeemWinningTicket(ticket.Hash())
+	if err == nil {
+		t.Error("expected zero deposit and penalty escrow error")
+	}
+	if err != nil && err != errZeroDepositAndPenaltyEscrow {
+		t.Errorf("expected zero deposit and penalty escrow error, got %v", err)
+	}
+
+	// Test zero deposit and non-zero penalty escrow
+	// Config stub broker with zero deposit and non-zero penalty escrow
+	b.SetDeposit(sender, big.NewInt(0))
+	b.SetPenaltyEscrow(sender, big.NewInt(500))
+
+	if err := r.RedeemWinningTicket(ticket.Hash()); err != nil {
+		t.Error(err)
+	}
+
+	used, err := b.IsUsedTicket(ticket)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !used {
+		t.Errorf("expected used ticket for sender with zero deposit and non-zero penalty escrow")
+	}
+
+	// Test non-zero deposit and zero penalty escrow
+	// Config stub broker with non-zero deposit and zero penalty escrow
+	b.SetDeposit(sender, big.NewInt(500))
+	b.SetPenaltyEscrow(sender, big.NewInt(0))
+
+	params2, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ticket = &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       0,
+		RecipientRandHash: params2.RecipientRandHash,
+	}
+
+	won, err = r.ReceiveTicket(ticket, sig, params2.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fatal()
+	}
+
+	if err := r.RedeemWinningTicket(ticket.Hash()); err != nil {
+		t.Error(err)
+	}
+
+	used, err = b.IsUsedTicket(ticket)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !used {
+		t.Errorf("expected used ticket for sender with non-zero deopsit and zero penalty escrow")
+	}
+
+	// Test invalid recipientRand revealed error
+	// This should fail because we use the same ticket with a recipientRand
+	// already invalidated
+	err = r.RedeemWinningTicket(ticket.Hash())
+	if err == nil {
+		t.Error("expected invalid recipientRand revealed error")
+	}
+	if err != nil && err != errInvalidRecipientRandRevealed {
+		t.Errorf("expected invalid recipientRand revealed error, got %v", err)
+	}
+
+	// TODO: Unit test for internal clearing of senderNonce memory?
```

Comment:
I agree with adding tests for memory cleanup. I also think we should test for when `r.broker.RedeemWinningTicket()` return an error.

In case of error from `broker.RedeemWinningTicket` I wonder if it makes sense that we still invalidate the `recipientRand` in question, versus perhaps only invalidating upon successful txn submission?

Another question is around retries. Should we expand what we're doing upon failure to redeem a winning ticket?

Reply:
I think the `recipientRand` should be invalidated as soon as the transaction is submitted (but the recipient should not wait until the transaction is confirmed since `recipientRand` is already visible in the mempool).

I imagined `broker.RedeemWinningTicket` to handle retries when the original transaction has an insufficient gas price. It can kickoff a goroutine for transaction retries and return an error indicating whether the original transaction was successfully submitted or not. Then, in `Recipient`, we would check the error returned and invalidate `recipientRand` if the error is nil.

Reply:
Sounds reasonable! IIUC the current implementation assumes that the txn would be submitted in the same thread as `recipient. RedeemWinningTicket`, right? - otherwise the `err` return value has no meaning. 

Sounds like you're suggesting to change this paradigm to actually submitting the txn in a separate goroutine?

Reply:
Yeah the original assumption was that `broker.RedeemWinningTicket` would run in the same thread and would block until the transaction was mined (and it would be ok to block here because `recipient.RedeemWinningTicket` would likely be run in its own thread). The error returned in this paradigm would inform the caller if the transaction was successfully mined or if the transaction retry process was unsuccessful (i.e. time out or exceeded gas price threshold while replacing transactions).

In the new paradigm that I am suggesting, `broker.RedeemWinningTicket` would:

1. Submit the transaction
2. Start a separate goroutine to wait for the transaction to be mined (and retry if necessary)
3. Return

If there is a problem with step 1, `broker.RedeemWinningTicket` would return an error immediately - since the transaction never made it to the network we wouldn't invalidate `recipientRand` here. If there is no problem with step 1, then nil is returned in step 3 so we invalidate `recipientRand` because we assume that the transaction made it to the network.

Reply:
SGTM!